### PR TITLE
adapter,controller: remove AllowWrites command

### PR DESCRIPTION
**Pull request by benesch**
@aljoscha @petrosagg putting this up to see how it fares in CI. We don't necessarily have to merge this PR if we think it's worth keeping the partial `AllowWrites` implementation around for optionality, but I found it much easier to reason about 0dt for sources without having to worry about `AllowWrites` (see https://github.com/MaterializeInc/materialize/pull/29699).

----

As our 0dt deployment thinking has evolved, we no longer plan to support transitioning a running `environmentd` process from read-only mode to read-write mode via an externally issued `AllowWrites` command. Instead, to facilitate the transition, we stop the read-only mode environmentd process and start a fresh environmentd process in read-write mode.

So, this commit removes the partial implementation of `AllowWrites` at the adapter/controller level. This will simplify the forthcoming implementation of 0dt for upsert sources, as we won't need the storage controller to keep track of which ingestions were started in read-only mode and which were not.

Touches https://github.com/MaterializeInc/materialize/issues/27413.

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/controller/src/lib.rs [hunk 1]
```diff
@@ -742,15 +714,8 @@ where
...
-        // mode. This way we keep that logic in one place.
-        if !read_only {
-            // We did not yet tell the controller about any collections, so
-            // there are no tables to initialize yet. Hence we don't need a
-            // register_ts.
-            let register_ts = None;
-            this.allow_writes(register_ts).await;
+        if !this.read_only {
+            this.remove_past_generation_replicas_in_background();
         }
```
Comment (teskje) [inline src/controller/src/lib.rs:719]: Nit: We could move this part before the creation of `Self` and then get rid of the `this` variable.

**Review by teskje — APPROVED** [review]
Controller changes LGMT. I think we could avoid doing some work in the storage collection write tasks in read-only mode, like only draining received updates and logging errors when they are unexpected, but I understand if you don't want to rock the boat too much here.

I'm in support of ripping out code we don't need!

File: src/controller/src/lib.rs (hunk 1 above)
Comment (benesch) [reply src/controller/src/lib.rs:719]: Well it's a method on `Self` 😅. I could inline the method in this method if you prefer though?

Comment (teskje) [reply src/controller/src/lib.rs:719]: 🤦 Yeah, that's a pretty good reason. No, I prefer the current version over inlining!

**Pull request merged**